### PR TITLE
FYLE-2cqxfzg: Unlink corporate card expenses which are implicitly merged

### PR DESCRIPTION
**Pull request by Dimple16**
Screenshot 2022-05-17 at 2 43 09 PM
Screenshot 2022-05-17 at 2 52 57 PM
Screenshot 2022-05-17 at 4 52 03 PM

### CODE REVIEW

File: src/app/core/services/transaction.service.ts [hunk 1]
```diff
@@ -744,4 +744,20 @@ export class TransactionService {
 
     return dialogBody;
   }
+
+  getUnlinkDialogBody() {
```
Comment (rvab) [inline src/app/core/services/transaction.service.ts:748]: Are we not mentioning return `type`?

Comment (rvab) [inline src/app/core/services/transaction.service.ts:748]: Are we not mentioning return type?

File: src/app/core/services/transaction.service.ts [hunk 2]
```diff
@@ -744,4 +744,20 @@ export class TransactionService {
...
+    </ul>`;
+
+    return dialogBody;
+  }
+
+  unlinkCorporateCardExpense(txnId: string) {
+    const data = {
+      txn_id: txnId,
+    };
+    return this.apiService.post('/transactions/undo_implicit_merge', data);
```
Comment (rvab) [inline src/app/core/services/transaction.service.ts:761]: API return `data type`?

File: src/app/fyle/add-edit-expense/add-edit-expense.page.ts [hunk 3]
```diff
@@ -829,6 +833,58 @@ export class AddEditExpensePage implements OnInit {
...
+        Type: 'unlink corporate card expense',
+        transaction: txnDetails?.tx,
+      };
+      this.trackingService.unlinkCorporateCardExpense(properties);
+      if (txnDetails?.tx?.report_id) {
+        this.router.navigate([
+          '/',
+          'enterprise',
+          'my_view_report',
+          { id: txnDetails?.tx?.report_id, navigateBack: true },
```
Comment (rvab) [inline src/app/fyle/add-edit-expense/add-edit-expense.page.ts:871]: `?` is not necessary `txnDetails?.tx?.report_id`

File: src/app/fyle/add-edit-expense/add-edit-expense.page.ts [hunk 4]
```diff
@@ -2840,6 +2908,8 @@ export class AddEditExpensePage implements OnInit {
       this.isCccExpense = etxn?.tx?.corporate_credit_card_expense_group_id;
       this.isExpenseMatchedForDebitCCCE = !!etxn?.tx?.corporate_credit_card_expense_group_id && etxn.tx.amount > 0;
       this.canDismissCCCE = !!etxn?.tx?.corporate_credit_card_expense_group_id && etxn.tx.amount < 0;
+      this.canUnlinkCCCE =
+        etxn?.tx?.is_implicitly_merged && ['APPROVER_PENDING', 'COMPLETE', 'DRAFT'].indexOf(etxn?.tx?.state) > -1;
```
Comment (rvab) [inline src/app/fyle/add-edit-expense/add-edit-expense.page.ts:2912]: use `includes` instead of `indexOf`

**Review by rvab — CHANGES_REQUESTED** [review]

File: src/app/fyle/add-edit-expense/add-edit-expense.page.ts [hunk 5]
```diff
@@ -829,6 +833,58 @@ export class AddEditExpensePage implements OnInit {
...
+      }
+      const toastMessage = 'Successfully removed the card details from the expense.';
+      const toastMessageData = {
+        message: toastMessage,
+      };
+      this.matSnackBar.openFromComponent(ToastMessageComponent, {
+        ...this.snackbarProperties.setSnackbarProperties('information', toastMessageData),
+        panelClass: ['msb-info'],
+      });
+      this.trackingService.showToastMessage({ ToastContent: toastMessageData.message });
```
[the remark this answers is not in the document]
Comment (Dimple16) [reply src/app/fyle/add-edit-expense/add-edit-expense.page.ts:884]: this was to gather data around toasts when it was newly added

File: src/app/core/models/v1/transaction.model.ts [hunk 6]
```diff
@@ -0,0 +1,77 @@
+import { List } from 'lodash';
```
Comment (rvab) [inline src/app/core/models/v1/transaction.model.ts:1]: Looks like unused.

**Review by rvab — APPROVED** [review]
Minor comment

**Pull request merged**